### PR TITLE
Add UPnP support (port forwarding, querying external IP)

### LINKED ISSUE #2970: Peer-to-Peer across NAT
Good article:
http://www.brynosaurus.com/pub/net/p2pnat/

Thanks

### PR DESCRIPTION
**Pull request by mhilbrunner**
This PR adds support for Universal Plug and Play (UPnP) to Godot's multiplayer capabilities.

This allows:
1. To forward ports automatically when hosting games behind a router, instead of requiring the user to do it themselves
2. To query the router for the external IP address

This is based on the BSD-licensed [miniupnpc](https://github.com/miniupnp/miniupnp/tree/master/miniupnpc). The only change to the upstream files was an additional header file that would usually be autogenerated with `cmake`. In our case, I added it manually.

**Changes**
- Adds *miniupnc* under `thirdparty/miniupnpc`
- Adds Godot implementation under `modules/upnp`
- Adds settings for module to scons

**Simplest example**
```
# Forward external UDP port 1234 on the router to us
var upnp = UPNP.new()
upnp.discover()
upnp.add_port_mapping(1234)
# Do something with the opened port, then on quitting:
upnp.delete_port_mapping(1234)
```

This autodetects the router and creates (or overwrites) the port mapping.
You can also customize the UPnP discovery params (by calling `upnp.discover()` yourself) and the way the gateway is chosen by using the methods on `UPNP` to iterate the `UPNPDevice`s. Once you found the device of your choosing, you can then call `add_port_mapping()` and `delete_port_mapping()` on that device, again providing the parameters of your choice. The methods on `UPNP` are just shortcuts. :smile: 

Of course, you can fine tune all relevant settings (like timeouts/ttl, internal and external ports, TCP vs UDP, lease time, description for the port forward etc.) from GDScript. 

**Glossary**
**IGD**: Internet Gateway Device, the UPnP type for routers
**NAT**: Network address translation
**UPnP**: Universal Plug and Play, the protocol

**Notes**
Not all routers are UPnP compatible. Not all devices support wildcards in port mappings, or a non-infinite duration for the lease, or a description. [Look at this list](http://miniupnp.free.fr/devicelist.php) to get an idea of how varied UPnP support is. The default parameters as they are now should maximize compatibility.

Relevant to https://github.com/godotengine/godot/issues/13947.
Closes https://github.com/godotengine/godot/issues/2970.

### CODE REVIEW

**Comment by akien-mga** [comment]
It fails building on Linux and macOS:
```
modules/upnp/upnp.cpp:115:38: error: use of undeclared identifier 'malloc'; did you mean 'alloca'?
        struct UPNPUrls* urls = (UPNPUrls *)malloc(sizeof(struct UPNPUrls));
                                            ^~~~~~
                                            alloca
/usr/include/alloca.h:32:14: note: 'alloca' declared here
extern void *alloca (size_t __size) __THROW;
             ^
modules/upnp/upnp.cpp:122:38: error: use of undeclared identifier 'malloc'; did you mean 'alloca'?
        struct IGDdatas* data = (IGDdatas *)malloc(sizeof(struct IGDdatas));
                                            ^~~~~~
                                            alloca
/usr/include/alloca.h:32:14: note: 'alloca' declared here
extern void *alloca (size_t __size) __THROW;
             ^
modules/upnp/upnp.cpp:133:2: error: use of undeclared identifier 'free'; did you mean 'DefaultAllocator::free'?
        free(xml);
        ^~~~
        DefaultAllocator::free
core/os/memory.h:68:29: note: 'DefaultAllocator::free' declared here
        _FORCE_INLINE_ static void free(void *p_ptr) { return Memory::free_static(p_ptr, false); }
                                   ^
3 errors generated.
```
Shouldn't you use the Godot memory management methods (`memnew`, `memfree`, `memalloc`)?

**Comment by mhilbrunner** [comment]
Yes, I should. :) faless just pointed them out to me too.

**Comment by mhilbrunner** [comment]
@akien-mga Fixed the build, was missing an `include`. Can't use Godot's memory management functions as the memory in question is either allocated by the library or freed by it.

File: modules/upnp/upnp.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+#include <reference.h>
+
+class UPNP : public Reference {
+
+	GDCLASS(UPNP, Reference);
+
+private:
+	String discover_multicast_if;
+	int discover_local_port;
+	bool discover_ipv6;
```
Comment (Faless) [inline modules/upnp/upnp.h]: UPNP class is missing a contructor where variable are not properly initialized. This causes them to have random initial values causing apparently random failure.

Comment (mhilbrunner) [reply modules/upnp/upnp.h]: Fixed :)

**Comment by mhilbrunner** [comment]
Added docs.

Will see about adding support for PCP and NAT-PMP in a seperate PR (using the BSD-licensed https://github.com/libpcp/pcp).

Then I'd like to add a "PortForwarder" Godot node on top of these that manages port forwards.
You'd add that to your multiplayer/lobby scene, and when it initializes it would automatically forward the ports according to the nodes configuration done in the editor, then free the ports once the scene exits.
This node would also automatically handle threading.

File: modules/upnp/upnpdevice.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+	ERR_FAIL_COND_V(i != UPNPCOMMAND_SUCCESS, "");
+
+	return String(addr);
+}
+
+int UPNPDevice::add_port_mapping(int port, int port_internal, String desc, String proto, int duration) const {
+	ERR_FAIL_COND_V(!is_valid_gateway(), UPNP::UPNP_RESULT_INVALID_GATEWAY);
+	ERR_FAIL_COND_V(port < 1 || port > 65535, UPNP::UPNP_RESULT_INVALID_PORT);
+	ERR_FAIL_COND_V(port < 0 || port > 65535, UPNP::UPNP_RESULT_INVALID_PORT); // Needs to allow 0 because 0 signifies "use external port as internal port"
```
Comment (Faless) [inline modules/upnp/upnpdevice.cpp]: you are doing both `port < 1` and `port < 0` check. Did you want to check for `port_internal` instead :wink: ?

**Review by Faless — CHANGES_REQUESTED** [review]
Great job :1st_place_medal: !
There are 2 small issues to fix (see comments).
And the question about javascript is open.

I've tested builds with my usual library schedule (X11-Linux/Server-Linux/Server-OpenBSD/Server-FreeBSD/OSX/iPhone/Android/Win/UWP/Javascript), and they all compile fine.
Also, port mapping seems to okay with my home router (had to enable IGDevice on it because it was factory disabled), as even though I can't see them in the usual router UI the port **does get open**.

Thank you for this PR! It is almost ready to merge :)

**Review by Faless — APPROVED** [review]

**Comment by akien-mga** [comment]
Only thing missing was to link against system miniupnpc on linuxbsd if `builtin_upnpc=no`, but I'll add that :)

**Comment by akien-mga** [comment]
And also missing adding the license details to `COPYRIGHT.txt`. I guess we should add this CODEOWNERS file and put me for `thirdparty/*` ;)

**Comment by tree786** [comment]
> Added docs.
> 
> Will see about adding support for PCP and NAT-PMP in a seperate PR (using the BSD-licensed https://github.com/libpcp/pcp).
> 
> Then I'd like to add a "PortForwarder" Godot node on top of these that manages port forwards.
> You'd add that to your multiplayer/lobby scene and configure the ports you wish to forward.
> It would then automatically forward those when the scene is run and remove the mapping on exit, handle threading and use PCP/NAT-PMP and fallback to UPnP if those didn't work.
> 
> You can of course also use the UPNP/UPNPDevice classes directly, if you want to do something custom. You just need to handle threading yourself or live with blocking, which may be fine if done at startup/shutdown or during loading.
> A little like HTTPRequest vs HTTPClient.

Are there currently any plans of adding a "PortForwarder" node for version 3.2 (or 4.0)?

**Pull request merged**